QgsFeatureSource

nyalldawson opened the pull request:
Adds a new interface QgsFeatureSource as an interface for classes which provide feature iterators via a getFeatures method

So far only QgsVectorLayer/QgsVectorDataProvider are QgsFeatureSources

other candidates could be:
- QgsVectorLayerCache. Tricky bit here is the non-const getFeatures, which requires either some ugly const_casting to avoid
- QgsFeatureStore. We'd need to copy a bunch of the memory layer iterator code here to create an iterator for the stored features.
- QgsAbstractFeatureSource. Again, getFeatures is non const here, which makes things tricky

So the big question here is whether QgsFeatureSource::getFeatures() *should* be const or non const. I can see benefits both ways. I'd imagine it's possible that some sources may have validly non-const getFeatures (e.g. some sort of temporary queue which drops features as they are retrieved). But making the interface non-const also makes things very messy and blocks us from ever using const QgsFeatureSources.

Opinions welcome. I'd personally lean toward const getFeatures().

nyalldawson committed:
Add QgsFeatureSource interface as interface for classes which
provide feature iterators via a getFeatures method

nyalldawson committed:
QgsVectorLayer, QgsVectorDataProvider are QgsFeatureSources

nyalldawson committed:
Add crs, fields, wkbType and featureCount to QgsFeatureSource

nyalldawson committed:
QgsFeatureSource/Sink can be stored in QVariant

nyalldawson committed:
Add missing overrides

m-kuhn commented:
Nice.
I'd say it should be const.
If something is non-const (temporary queue etc.) it shouldn't be passed as a QgsFeatureSource. In such cases iterators accessing this source should be passed around. This also allows for controlling when we are able to pop features from the queue (because they have been served on all iterators).

wonder-sk reviewed (approved):
Also +1 for const getFeatures()

nyalldawson committed:
Split ProviderTestCase into provider/source tests

There's now a FeatureSourceTestCase base class which handles
testing QgsFeatureSources for conformance. These tests
consist mostly of tests which where previously in
ProviderTestCase. ProviderTestCase now subclasses
FeatureSourceTestCase + adds tests which are specific
to QgsVectorDataProviders.

nyalldawson committed:
Fix some providers not requesting all required attributes needed
for client side order by clauses
